plugins/support/lib/nokul/support/rest_client.rb

```diff
@@ -81,14 +83,16 @@ def code
           http_response.code.to_i
         end
 
+        def decode
+          body && JSON.parse(body, symbolize_names: true)
+        end
+
         def error!
-          http_response.error! unless code.between?(200, 299)
+          http_response.error! unless success?
         end
 
-        def unmarshal_json
-          JSON.parse(body || '')
-        rescue JSON::ParserError
-          raise UnmarshalJSONError, 'response body is not parseable'
+        def success?
+          code.between?(200, 299)
```

Comment:
Buna gerek yok, response nesnesinin sınıfına bakman yeterli, örneğin `Net::HTTPSuccess` ise geçerli tüm `2xx` kodlarını kapsar.  Kanonik örneği takip edin:

```ruby
  response = Net::HTTP.get_response(url)

  case response
  when Net::HTTPSuccess then
    response
  ...
```

Dolayısıyla şöyle olmalı:

```ruby
def success?
  http_response.kind_of? NetHTTPSuccess # is_a? de olur, alias ikisi
end
```

Fakat...  Soru şu?  Rest client için Tüm 2xx kodlarını niye kabul ediyoruz, kabul etmeli miyiz?  Bu kodların bir kısmı POST/PUT dönüşleri için anlamlı, bir kısmında, örneğin `204` body yok (`nil`).[1]  Bana kalırsa sadece `NetHTTPOK` kullanılsın (ama kararımızı yine `success?` içinde ifade edelim).

```ruby
def success?
  http_response.kind_of? NetHTTPOK
end
```

Bu eğer genel amaçlı bir HTTP grabber olsaydı böyle yapmazdık, fakat sadece Rest API isteği çıkacağız.

[1]: Bazı kaynaklar:
  - https://www.koozai.com/blog/analytics/server-response-codes-and-what-they-mean/
  -  https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
  Deneme yapmak için https://httpstat.us servisini kullanabilirsiniz.

Reply:
Hocam burada 2XX'ler success kabul edilmeliyiz, body'nin `nil` dönme durumu bir çok senaryoda olabilir. Örneğin; kişinin YÖKSİS'te kaydı yoksa body `nil' dönüyor. Yani işlem başarılı content yok.  Eğer 204'ü hata kabul edersek diğer tarafta hata yönetimi yapmak gerekecek. Bu pekte istenilen bir durum değil.

Reply:
Tamamdır `NetHTTPSuccess` üzerinden gidelim o halde.  Kafama takılan [şu](https://github.com/omu/nokul/blob/448fdcbb193da2bd221254ed37cc7791a313af3a/app/services/xokul/connection.rb#L15): `body && JSON.parse` yaptığımızda bu `nil` değeri yukarı iletiliyor ve `Connection.request` üzerinden daha da yukarı iletiliyor.  Bu nasıl yönetiliyor?  Hızlıca baktığımda şu an yönetilmediğini görüyorum.  Önceden hatalı yönetiliyordu (`body || ''` ile JSON parse hatası üretmek).  Bunu ne yapacağız, karar verelim.

Reply:
Hocam `nil` durum [*response] şeklinde yönetiliyor ve eğer response nil ise boş array olarak işleme alınıyor.

Reply:
@roktas hocam `nil`i yönetmemiz gerekmiyor bence. eğer body `nil` ise (bu çoğu zaman `204` kodlu response'da oluyor) en tepede çağırana kadar `nil` döner. eğer body boş değil ise parse etmeyi dener. parse edebiliyorsa çağırana parse edilmiş veriyi döner, edemiyorsa log düşüp yoluna devam eder. nil ile çağıran ilgileniyor. öncekinde neden öyle yaptığımı hatırlamıyorum. sanırım `JSON.parse()` nil kabul etmiyordu ve bir workaround ile çözüp geçmişim :thinking:

Reply:
@huseyin her yerde "conveniency" olarak `Connection.request` kullanılıyor.  @isubas'ın not ettiği gibi mevcut durumda zaten bir şekilde yönetiliyor, ki öyle olmalı.  Bunlar uygulamanın dış dünyayla kontak noktaları.  Kontak sonrası ne beklendiği önemli.  Exception'lar içte yönetiliyor ama bir metod ayrıca nil dönebiliyorsa farkındalık oluşturmak için bunu mutlaka dokümante etmek lazım.